### PR TITLE
[WIP] Sexy Notifications (ie. Notifications Revamp)

### LINKED ISSUE #849: UI: Notification experience improvements
@jordwalke had some great ideas around improving our notifications story in the Discord.

A persistent number of 'new messages' that are 'unread' in the statusbar. 
image

Message animates in when a new notification comes in:
image

Messages are shown in the bottom:
image

And can expand the pane:
image

### PR DESCRIPTION
**Pull request by glennsl**
Fixes #849

Peek 2020-01-08 05-26

This can be merged as-is, although there are still some issues and remaining work:

- [x] Pass along background color to all text elements in order to get proper gamma correction
- [ ] Context menu with Clear All action (turned out to be more pain to implement than expected, for various reasons)
- [ ] Animations will replay when older notifications are removed since keys aren't actually used during reconciliation (needs to be fixed upstream in Revery).
- [ ] Garbage collect expired notifications in the `useExpiration` hook.
- [ ] Refactor Pane modules

### CODE REVIEW

File: src/Core/Utility.re [hunk 1]
```diff
@@ -894,3 +894,15 @@ module Json = {
...
+module ColorEx = {
+  open Revery;
+
+  let mix = (~start, ~stop, ~amount) =>
+    Color.{
+      r: (stop.r -. start.r) *. amount +. start.r,
+      g: (stop.g -. start.g) *. amount +. start.g,
+      b: (stop.b -. start.b) *. amount +. start.b,
+      a: (stop.a -. start.a) *. amount +. start.a,
+    };
```
Comment (glennsl) [inline src/Core/Utility.re]: Should probably be upstreamed to Revery

File: src/UI/GlobalContext.re [hunk 2]
```diff
@@ -24,7 +24,7 @@ type t = {
   setActiveWindow: (int, int) => unit,
   openEditorById: int => unit,
   closeEditorById: int => unit,
-  hideNotification: int => unit,
+  hideNotification: Notification.t => unit,
```
Comment (glennsl) [inline src/UI/GlobalContext.re]: Should be removed in favor of using `dispatch`

File: src/UI/StatusBar.re [hunk 3]
```diff
@@ -14,32 +14,193 @@ open Oni_Model;
...
+          |> List.filter(item =>
+               !List.exists(((it, _t)) => equals(it, item), active)
+             );
+
+        if (untracked != []) {
+          let init = item => (item, time);
+          setActive(tracked => List.map(init, untracked) @ tracked);
+        };
+
+        // TODO: Garbage collection of expired, but on what condition?
```
Comment (glennsl) [inline src/UI/StatusBar.re]: TODO

**Comment by glennsl** [comment]
I've tried to get a context menu working, but am bumping up against a lot of rough edges. First with right-click not working (fixed in https://github.com/revery-ui/revery/pull/706), then a bunch of issues with value restriction errors caused by the brisk ppx, and now nested clickables not working.

I've therefore moved it [a separate branch], although still based on this, and recommend this be merged as-is since it might take a while to sort out these issues as well as the design of it.

**Comment by bryphe** [comment]
Wow! 

This looks incredible @glennsl - such a huge improvement over what we had before. Awesome work!

File: src/Core/Utility.re (hunk 1 above)
Comment (bryphe) [reply src/Core/Utility.re]: Makes sense, this would be helpful in the `Color` module

**Comment by bryphe** [comment]
> I've therefore moved it a separate branch, although still based on this, and recommend this be merged as-is since it might take a while to sort out these issues as well as the design of it.

Sounds good - thanks for fixing some of the initial issues (like the mouse click issue) - tracking the context menu in a separate branch sounds reasonable 👍 The notification experience w/o the context menu is still so much improved over what we have today.

File: src/Model/Notification.re [hunk 4]
```diff
@@ -1,12 +1,29 @@
-open Actions;
+[@deriving show({with_path: false})]
+type t = {
```
Comment (bryphe) [inline src/Model/Notification.re]: Thanks for cleaning up the types here and decoupling from Actions

File: src/UI/CustomHooks.re [hunk 5]
```diff
@@ -0,0 +1,41 @@
+open Oni_Core;
+
+open Revery;
+open Revery.UI;
+
+module ColorEx = Utility.ColorEx;
+
+let colorTransition =
```
Comment (bryphe) [inline src/UI/CustomHooks.re]: Neat way to build on the `animation` hook 👍

Comment (glennsl) [reply src/UI/CustomHooks.re]: Should we upstream this to Revery too? It seems a little bit special-purpose, and should perhaps ideally be in some extended library for Revery.

**Pull request merged**